File: LeetCode/Patterns/Dynamic Programming/152. Maximum Product Subarray/Solution.go
func maxProduct(nums []int) int {
	n := len(nums)
	prefix, suffix, maxProduct := 1, 1, nums[0]
	for i := 0; i < n; i++ {
		j := n - i - 1
		if prefix == 0 {
			prefix = 1
		}
		if suffix == 0 {
			suffix = 1
		}
		prefix *= nums[i]
		suffix *= nums[j]
		maxProduct = max(maxProduct, max(prefix, suffix))
	}
	return maxProduct
}

File: LeetCode/Patterns/Dynamic Programming/152. Maximum Product Subarray/Solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int size = nums.length;
        if (size == 1) {
            return nums[0];
        }
        int prefix = 1;
        int suffix = 1;
        int result = Integer.MIN_VALUE;
        for (int i = 0; i < size; i++) {
            if (prefix == 0) {
                prefix = 1;
            }
            if (suffix == 0) {
                suffix = 1;
            }
            prefix *= nums[i];
            suffix *= nums[size - i - 1];
            result = Math.max(result, Math.max(prefix, suffix));
        }
        return result;
    }
}